Trigger/src/EvalWeightedEventCounts_module.cc

```diff
@@ -0,0 +1,223 @@
+
+private:
+  int _diagLevel;
+  art::InputTag _evtWeightTag;
+
+  float _nProcess;
+  std::string _tagName;
+  double _nPOT;
+
+  occupancyHist_ _occupancyHist;
+
+  const mu2e::Tracker* _tracker;
+
+  // the following pointer is needed to navigate the MC truth info of the strawHits
+  const art::Event* _event;
+
+  float _minPOT, _maxPOT;
+  double _wgSum[3] = {0};
+};
+
+EvalWeightedEventCounts::EvalWeightedEventCounts(const Parameters& config) :
+    art::EDAnalyzer(config), _diagLevel(config().diagLevel()), _evtWeightTag(config().evtWeightTag()),
+    _nProcess(config().nProcess()), _tagName(config().tagName()), _minPOT(1e6), _maxPOT(4e8) {}
+
+void EvalWeightedEventCounts::bookHistograms() {
+  art::ServiceHandle<art::TFileService> tfs;
+
+  bookOccupancyInfoHist(tfs, _occupancyHist);
+}
+
+void EvalWeightedEventCounts::bookOccupancyInfoHist(art::ServiceHandle<art::TFileService>& Tfs,
+                                                    occupancyHist_& Hist) {
+
+  int index_last = 0;
+  art::TFileDirectory occInfoDir = Tfs->mkdir("occInfoGeneral");
+  Hist._hOccInfo[index_last][0] = occInfoDir.make<TH1F>(
+      Form("hInstLum_%i", index_last), "distrbution of instantaneous lum; p/#mu-bunch", 1000,
+      _minPOT, _maxPOT);
+  Hist._hOccInfo[index_last][1] = occInfoDir.make<TH1F>(
+      Form("hInstLum2B_%i", index_last),
+      "distrbution of instantaneous lum re-weighted in 2 batch-mode; p/#mu-bunch", 1000, _minPOT,
+      _maxPOT);
+}
+
+//--------------------------------------------------------------------------------//
+void EvalWeightedEventCounts::beginJob() { bookHistograms(); }
+
+//--------------------------------------------------------------------------------//
+void EvalWeightedEventCounts::endJob() {
+
+  //    evalTriggerRate();
+  printf("[%s::enJob] totWg    = %10.3e\n", _tagName.c_str(), _wgSum[0]);
+  printf("[%s::enJob] totEvtB1 = %10.3f\n", _tagName.c_str(), _wgSum[1]);
+  printf("[%s::enJob] totEvtB2 = %10.3f\n", _tagName.c_str(), _wgSum[2]);
+}
+
+//--------------------------------------------------------------------------------
+
+//================================================================
+void EvalWeightedEventCounts::beginRun(const art::Run& run) {
+  // get bfield
+  // GeomHandle<BFieldManager> bfmgr;
+  // GeomHandle<DetectorSystem> det;
+  // CLHEP::Hep3Vector vpoint_mu2e = det->toMu2e(CLHEP::Hep3Vector(0.0,0.0,0.0));
+  // _bz0 = bfmgr->getBField(vpoint_mu2e).z();
+
+  // mu2e::GeomHandle<mu2e::Tracker> th;
+  // _tracker  = th.get();
+}
+
+//--------------------------------------------------------------------------------
+void EvalWeightedEventCounts::analyze(const art::Event& event) {
+
+  // get the number of POT
+  double lumi = -1.;
+  art::Handle<ProtonBunchIntensity> evtWeightH;
+  event.getByLabel(_evtWeightTag, evtWeightH);
+  if (evtWeightH.isValid()) {
+    lumi = (double)evtWeightH->intensity();
+  }
+
```

Comment:
I don't like what the code will do if the handle is invalid.  What does the call to ROOT::Math::lognormal_cdf do when the first argument is -1?  Also, the code will give negative contributions to the histograms and the sums, which will silently produce misleading results.

Generally it's a bad idea to make up something when you have invalid input.  Instead you should indicate invalid input and either skip the event or throw an exception.    You should speak with Giani to learn what he wants.

Reply:
Fixed the issue.

Reply:
One more fix here.  Remove line 140 and the type to line 142.

Also, we do not want people to use c-style casts.  The C++ automatic type conversions will do the right thing without any cast.  If a cast is needed choose an appropriate C++ cast; here static_cast would be OK but it is not needed.

Reply:
Fixed it!

Reply:
Fixed it.